### ISSUE TITLE
SSEServerTransport does not work with bun runtime

### ISSUE THREAD

**Issue by the-vampiire** [modelcontextprotocol/typescript-sdk#201]
**Describe the bug**
the bun runtime does not work with the `SSEServerTransport` in a typescript project. 

there are no errors reported but it appears the SSE connection does not open successfully. 

after many hours trying to debug this using the MCP debugger i realized it works fine when running with `npx tsx`. this indicates an issue with the bun runtime as `npx tsx` uses the nodejs runtime (with the esbuild integration for transpiling).

**To Reproduce**
1. create a basic MCP SSE server ([minimal repo link](https://github.com/the-vampiire/mcp-sse))
2. run using bun 
3. run the MCP inspector `npx @modelcontextprotocol/inspector`
4. open the browser to the inspector `http://localhost:5173`
5. open dev tools network tab
6. select SSE transport and enter URL `http://localhost:3500/sse`
7. try to connect
8. observe inspector behavior and infinitely pending network call to `/sse` endpoint
9. repeat using `npx tsx`
10. no issues

**Expected behavior**
i expected to connect to the MCP SSE server running in the bun runtime

**Logs**

## bun
> MCP SSE server logs

connection never succeeds (SSE connection pending, no messages exchanged), no other options are available

```sh
bun sse.ts
listening on port 3500
connecting to transport c17254d3-5703-4de4-b9d8-df0cd2d3c604
<end, repeats on each "connect" button click>
```

> MCP inspector logs

```sh
npx @modelcontextprotocol/inspector
Starting MCP inspector...
Proxy server listening on port 3000

🔍 MCP Inspector is up and running at http://localhost:5173 🚀
New SSE connection
Query parameters: { transportType: 'sse', url: 'http://localhost:3500/sse' }
SSE transport: url=http://localhost:3500/sse, headers=
<end, repeats on each "connect" button click>
```

## node

connection succeeds, list tools etc all work fine

> MCP SSE server logs

```sh
npx tsx sse.ts
listening on port 3500
connecting to transport f51cda32-ddab-4263-9a96-42b210bc31ff
received message, checking for transport
posting message to transport f51cda32-ddab-4263-9a96-42b210bc31ff
received message, checking for transport
posting message to transport f51cda32-ddab-4263-9a96-42b210bc31ff
```

> MCP inspector logs

```sh
npx @modelcontextprotocol/inspector
Starting MCP inspector...
Proxy server listening on port 3000

🔍 MCP Inspector is up and running at http://localhost:5173 🚀
New SSE connection
Query parameters: { transportType: 'sse', url: 'http://localhost:3500/sse' }
SSE transport: url=http://localhost:3500/sse, headers=
Connected to SSE transport
Connected MCP client to backing server transport
Created web app transport
Created web app transport
Set up MCP proxy
Received message for sessionId dfba9df5-53c8-4d27-abc6-8c28d1db009c
Received message for sessionId dfba9df5-53c8-4d27-abc6-8c28d1db009c
```

**Additional context**

i also had this issue with claude code and cursor which led me to using the inspector

**Comment by the-vampiire** [comment]
related:
- https://github.com/oven-sh/bun/issues/16562
- https://github.com/oven-sh/bun/issues/2443

i was unaware of the bun SSE limitation (looking into it more now). unclear if this is something that can be fixed in the SDK code or if it needs upstream support from bun. let me know if its something i can help with.

**Comment by malangcat** [comment]
https://github.com/oven-sh/bun/issues/15574

Bun currently has a behavior on Windows where the first SSE message isn't sent to the client until a second message is written. For example, if your code writes only one SSE message, Bun will not send it immediately. Instead, it waits until a subsequent message is written before sending the first one.

For workaround, we might write dummy message before first endpoint event. https://github.com/modelcontextprotocol/typescript-sdk/blob/0fa2397174eba309b54575294d56754c52b13a65/src/server/sse.ts#L52

**Comment by malangcat** [comment]
Writing dummy event right after SSEServerTransport creation worked for me.

```
      this.sseTransport = new SSEServerTransport(
        "/messages",
        res as unknown as ServerResponse<IncomingMessage>,
      );
      res.write('event: log\ndata: "dummy event for bun workaround"\n\n');
```

**Comment by the-vampiire** [comment]
nice find man! I’m on macos but gave it a try. 

your solution didnt work because it expects a text/event-stream response and to come after the connect call. but it did give me a good clue and i found a solution:

```ts
app.get("/sse", async (req, res) => {
  transport = new SSEServerTransport("/message", res);
  console.log(`connecting to transport ${transport.sessionId}`);
  await server.connect(transport);
  await transport.send({
    jsonrpc: "2.0",
    id: Number(Date.now),
    result: { log: true },
  });
});
```

interestingly this dummy message was ignored in both the inspector and UI:

<img width="1683" alt="Image" src="https://github.com/user-attachments/assets/8412bda8-cd24-4f84-a661-1760cf8954a8" />

**Comment by the-vampiire** [comment]
thats weird. about 1 minute later i got a zod error in the inspector console. nothing reported in the UI and it is still working fine for listing / calling tools.

**Comment by the-vampiire** [comment]
modded the `sse` file and it works! nice find man.

```diff
    async start() {
        if (this._sseResponse) {
            throw new Error("SSEServerTransport already started! If using Server class, note that connect() calls start() automatically.");
        }
        this.res.writeHead(200, {
            "Content-Type": "text/event-stream",
            "Cache-Control": "no-cache",
            Connection: "keep-alive",
        });
        // Send the endpoint event
        this.res.write(`event: endpoint\ndata: ${encodeURI(this._endpoint)}?sessionId=${this._sessionId}\n\n`);

+        if (process.versions.bun) {
+            this.res.write('event: log\ndata: bun runtime compatibility message\n\n');
+        }
        this._sseResponse = this.res;
        this.res.on("close", () => {
            var _a;
            this._sseResponse = undefined;
            (_a = this.onclose) === null || _a === void 0 ? void 0 : _a.call(this);
        });
    }
```

**Comment by the-vampiire** [comment]
fixed in bun 1.2.6
